Debugger hangs

## Environment data
VS Code version:  1.13.1
Python Extension version:  0.6.5
Python Version:  3.5.2 
OS and version:  Ubuntu 16.04 LTS (running as a virtual machine in VMWare Player)

## Actual behavior
PyTest periodically hangs instead of running/debugging unit tests. In the case of debugging, rather than the status bar at the bottom of the screen turning orange, it remains blue and the "Running Tests" indicator seems to blink forever. Clicking it and selecting the "Stop running tests" command has no effect.

In order to proceed I have to close VS Code and open it again.

Here's a picture of the indicator bar when the bug happens:
![image](https://user-images.githubusercontent.com/8980400/27346568-7f512ab0-55a2-11e7-868a-200ce42d7cf3.png)

## Expected  behavior
The tests should run/debug as expected, or at least let me cancel them if they hang.

## Steps to reproduce:  
Hard to say as the problem is intermittent. I notice it most often when I try to debug a specific unit test by pressing the "Debug Test" button above the test definition.  Looking at the console log, the issue maybe be caused by a bug in "Jedi" (not sure what that is).

## Settings  
Your launch.json (if dealing with debugger issues):  
```{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Attach (Remote Debug)",
            "type": "python",
            "request": "attach",
            "localRoot": "${workspaceRoot}",
            "remoteRoot": "${workspaceRoot}",
            "port": 3000,
            "secret": "my_secret",
            "host": "localhost"
        },
        {
            "name": "Python",
            "type": "python",
            "request": "launch",
            "stopOnEntry": false,
            "pythonPath": "${config:python.pythonPath}",
            "program": "${file}",
            "cwd": "${workspaceRoot}",
            "env": {},
            "envFile": "${workspaceRoot}/.env",
            "debugOptions": [
                "WaitOnAbnormalExit",
                "WaitOnNormalExit",
                "RedirectOutput"
            ]
        },
        {
            "name": "PySpark",
            "type": "python",
            "request": "launch",
            "stopOnEntry": true,
            "osx": {
                "pythonPath": "${env:SPARK_HOME}/bin/spark-submit"
            },
            "windows": {
                "pythonPath": "${env:SPARK_HOME}/bin/spark-submit.cmd"
            },
            "linux": {
                "pythonPath": "${env:SPARK_HOME}/bin/spark-submit"
            },
            "program": "${file}",
            "cwd": "${workspaceRoot}",
            "env": {},
            "envFile": "${workspaceRoot}/.env",
            "debugOptions": [
                "WaitOnAbnormalExit",
                "WaitOnNormalExit",
                "RedirectOutput"
            ]
        },
        {
            "name": "Python Module",
            "type": "python",
            "request": "launch",
            "stopOnEntry": true,
            "pythonPath": "${config:python.pythonPath}",
            "module": "module.name",
            "cwd": "${workspaceRoot}",
            "env": {},
            "envFile": "${workspaceRoot}/.env",
            "debugOptions": [
                "WaitOnAbnormalExit",
                "WaitOnNormalExit",
                "RedirectOutput"
            ]
        },
        {
            "name": "Integrated Terminal/Console",
            "type": "python",
            "request": "launch",
            "stopOnEntry": true,
            "pythonPath": "${config:python.pythonPath}",
            "program": "${file}",
            "cwd": "",
            "console": "integratedTerminal",
            "env": {},
            "envFile": "${workspaceRoot}/.env",
            "debugOptions": [
                "WaitOnAbnormalExit",
                "WaitOnNormalExit"
            ]
        },
        {
            "name": "External Terminal/Console",
            "type": "python",
            "request": "launch",
            "stopOnEntry": true,
            "pythonPath": "${config:python.pythonPath}",
            "program": "${file}",
            "cwd": "",
            "console": "externalTerminal",
            "env": {},
            "envFile": "${workspaceRoot}/.env",
            "debugOptions": [
                "WaitOnAbnormalExit",
                "WaitOnNormalExit"
            ]
        },
        {
            "name": "Django",
            "type": "python",
            "request": "launch",
            "stopOnEntry": true,
            "pythonPath": "${config:python.pythonPath}",
            "program": "${workspaceRoot}/manage.py",
            "cwd": "${workspaceRoot}",
            "args": [
                "runserver",
                "--noreload"
            ],
            "env": {},
            "envFile": "${workspaceRoot}/.env",
            "debugOptions": [
                "WaitOnAbnormalExit",
                "WaitOnNormalExit",
                "RedirectOutput",
                "DjangoDebugging"
            ]
        },
        {
            "name": "Flask",
            "type": "python",
            "request": "launch",
            "stopOnEntry": false,
            "pythonPath": "${config:python.pythonPath}",
            "program": "fully qualified path fo 'flask' executable. Generally located along with python interpreter",
            "cwd": "${workspaceRoot}",
            "env": {
                "FLASK_APP": "${workspaceRoot}/quickstart/app.py"
            },
            "args": [
                "run",
                "--no-debugger",
                "--no-reload"
            ],
            "envFile": "${workspaceRoot}/.env",
            "debugOptions": [
                "WaitOnAbnormalExit",
                "WaitOnNormalExit",
                "RedirectOutput"
            ]
        },
        {
            "name": "Flask (old)",
            "type": "python",
            "request": "launch",
            "stopOnEntry": false,
            "pythonPath": "${config:python.pythonPath}",
            "program": "${workspaceRoot}/run.py",
            "cwd": "${workspaceRoot}",
            "args": [],
            "env": {},
            "envFile": "${workspaceRoot}/.env",
            "debugOptions": [
                "WaitOnAbnormalExit",
                "WaitOnNormalExit",
                "RedirectOutput"
            ]
        },
        {
            "name": "Pyramid",
            "type": "python",
            "request": "launch",
            "stopOnEntry": true,
            "pythonPath": "${config:python.pythonPath}",
            "cwd": "${workspaceRoot}",
            "env": {},
            "envFile": "${workspaceRoot}/.env",
            "args": [
                "${workspaceRoot}/development.ini"
            ],
            "debugOptions": [
                "WaitOnAbnormalExit",
                "WaitOnNormalExit",
                "RedirectOutput",
                "Pyramid"
            ]
        },
        {
            "name": "Watson",
            "type": "python",
            "request": "launch",
            "stopOnEntry": true,
            "pythonPath": "${config:python.pythonPath}",
            "program": "${workspaceRoot}/console.py",
            "cwd": "${workspaceRoot}",
            "args": [
                "dev",
                "runserver",
                "--noreload=True"
            ],
            "env": {},
            "envFile": "${workspaceRoot}/.env",
            "debugOptions": [
                "WaitOnAbnormalExit",
                "WaitOnNormalExit",
                "RedirectOutput"
            ]
        },
        {
            "name": "Attach (Remote Debug)",
            "type": "python",
            "request": "attach",
            "localRoot": "${workspaceRoot}",
            "remoteRoot": "${workspaceRoot}",
            "port": 3000,
            "secret": "my_secret",
            "host": "localhost"
        }
    ]
}
```   
Your settings.json:   
```{
    "python.unitTest.pyTestArgs": [
        "Tests"
    ],
    "python.unitTest.pyTestEnabled": true
}
```   
## Logs
Output from ```Python``` output panel  
None

Output from ```Console window``` (Help->Developer Tools menu)
```extensionHost.ts:293 [Extension Host] ---------------------------
d.logExtensionHostMessage @ extensionHost.ts:293
extensionHost.ts:293 [Extension Host] stderr jediProxy
d.logExtensionHostMessage @ extensionHost.ts:293
extensionHost.ts:293 [Extension Host] Error (stderr) Traceback (most recent call last):
  File "completion.py", line 588, in watch
    self._process_request(rq)
  File "completion.py", line 573, in _process_request
    request.get('prefix', '')))
  File "completion.py", line 196, in _serialize_completions
    completions = script.completions()
  File "preview/jedi/api/__init__.py", line 173, in completions
    completions = completion.completions()
  File "preview/jedi/api/completion.py", line 90, in completions
    completion_names = self._get_context_completions()
  File "preview/jedi/api/completion.py", line 118, in _get_context_completions
    grammar, self._code_lines, self._module, self._position
  File "preview/jedi/api/helpers.py", line 153, in get_stack_at_position
    raise SystemError("This really shouldn't happen. There's a bug in Jedi.")
SystemError: This really shouldn't happen. There's a bug in Jedi.


d.logExtensionHostMessage @ extensionHost.ts:293
extensionHost.ts:293 [Extension Host] ---------------------------
d.logExtensionHostMessage @ extensionHost.ts:293
extensionHost.ts:293 [Extension Host] stderr jediProxy
d.logExtensionHostMessage @ extensionHost.ts:293
extensionHost.ts:293 [Extension Host] Error (stderr) Traceback (most recent call last):
  File "completion.py", line 588, in watch
    self._process_request(rq)
  File "completion.py", line 573, in _process_request
    request.get('prefix', '')))
  File "completion.py", line 196, in _serialize_completions
    completions = script.completions()
  File "preview/jedi/api/__init__.py", line 173, in completions
    completions = completion.completions()
  File "preview/jedi/api/completion.py", line 90, in completions
    completion_names = self._get_context_completions()
  File "preview/jedi/api/completion.py", line 118, in _get_context_completions
    grammar, self._code_lines, self._module, self._position
  File "preview/jedi/api/helpers.py", line 153, in get_stack_at_position
    raise SystemError("This really shouldn't happen. There's a bug in Jedi.")
SystemError: This really shouldn't happen. There's a bug in Jedi.


d.logExtensionHostMessage @ extensionHost.ts:293
extensionHost.ts:293 [Extension Host] ---------------------------
d.logExtensionHostMessage @ extensionHost.ts:293
extensionHost.ts:293 [Extension Host] stderr jediProxy
d.logExtensionHostMessage @ extensionHost.ts:293
extensionHost.ts:293 [Extension Host] Error (stderr) Traceback (most recent call last):
  File "completion.py", line 588, in watch
    self._process_request(rq)
  File "completion.py", line 573, in _process_request
    request.get('prefix', '')))
  File "completion.py", line 196, in _serialize_completions
    completions = script.completions()
  File "preview/jedi/api/__init__.py", line 173, in completions
    completions = completion.completions()
  File "preview/jedi/api/completion.py", line 90, in completions
    completion_names = self._get_context_completions()
  File "preview/jedi/api/completion.py", line 118, in _get_context_completions
    grammar, self._code_lines, self._module, self._position
  File "preview/jedi/api/helpers.py", line 153, in get_stack_at_position
    raise SystemError("This really shouldn't happen. There's a bug in Jedi.")
SystemError: This really shouldn't happen. There's a bug in Jedi.


d.logExtensionHostMessage @ extensionHost.ts:293
index.js:41 [uncaught exception]: Error: read ECONNRESET
onError @ index.js:41
index.js:44 Error: read ECONNRESET
    at exports._errnoException (util.js:1022)
    at Pipe.onread (net.js:572)
onError @ index.js:44
events.js:160 Uncaught Error: read ECONNRESET
    at exports._errnoException (util.js:1022)
    at Pipe.onread (net.js:572)
index.js:41 [uncaught exception]: Error: read ECONNRESET
onError @ index.js:41
index.js:44 Error: read ECONNRESET
    at exports._errnoException (util.js:1022)
    at Pipe.onread (net.js:572)
onError @ index.js:44
events.js:160 Uncaught Error: read ECONNRESET
    at exports._errnoException (util.js:1022)
    at Pipe.onread (net.js:572)
index.js:41 [uncaught exception]: Error: read ECONNRESET
onError @ index.js:41
index.js:44 Error: read ECONNRESET
    at exports._errnoException (util.js:1022)
    at Pipe.onread (net.js:572)
onError @ index.js:44
events.js:160 Uncaught Error: read ECONNRESET
    at exports._errnoException (util.js:1022)
    at Pipe.onread (net.js:572)
index.js:41 [uncaught exception]: Error: read ECONNRESET
onError @ index.js:41
index.js:44 Error: read ECONNRESET
    at exports._errnoException (util.js:1022)
    at Pipe.onread (net.js:572)
onError @ index.js:44
events.js:160 Uncaught Error: read ECONNRESET
    at exports._errnoException (util.js:1022)
    at Pipe.onread (net.js:572)
extensionHost.ts:293 [Extension Host] TextEditor disposed
d.logExtensionHostMessage @ extensionHost.ts:293
extensionHost.ts:293 [Extension Host] TextEditor disposed
d.logExtensionHostMessage @ extensionHost.ts:293
extensionHost.ts:293 [Extension Host] TextEditor disposed
d.logExtensionHostMessage @ extensionHost.ts:293
extensionHost.ts:293 [Extension Host] TextEditor disposed
d.logExtensionHostMessage @ extensionHost.ts:293
extensionHost.ts:293 [Extension Host] TextEditor disposed
d.logExtensionHostMessage @ extensionHost.ts:293
```

--- a/src/client/unittests/common/baseTestManager.ts
+++ b/src/client/unittests/common/baseTestManager.ts
@@ -5,6 +5,11 @@ import { resetTestResults, displayTestErrorMessage, storeDiscoveredTests } from 
 import { Installer, Product } from '../../common/installer';
 import { isNotInstalledError } from '../../common/helpers';
 
+enum CancellationTokenType {
+    testDicovery,
+    testRunner
+}
+
 export abstract class BaseTestManager {
     private tests: Tests;
     private _status: TestStatus = TestStatus.Unknown;
@@ -49,19 +54,21 @@ export abstract class BaseTestManager {
 
         resetTestResults(this.tests);
     }
-    private createCancellationTokens(createTestRunnerToken: boolean = false) {
-        this.disposeCancellationTokens(createTestRunnerToken);
-        this.testDiscoveryCancellationTokenSource = new vscode.CancellationTokenSource();
-        if (createTestRunnerToken) {
+    private createCancellationToken(tokenType: CancellationTokenType) {
+        this.disposeCancellationToken(tokenType);
+        if (tokenType === CancellationTokenType.testDicovery) {
+            this.testDiscoveryCancellationTokenSource = new vscode.CancellationTokenSource();
+        } else {
             this.testRunnerCancellationTokenSource = new vscode.CancellationTokenSource();
         }
     }
-    private disposeCancellationTokens(disposeTestRunnerToken: boolean = false) {
-        if (this.testDiscoveryCancellationTokenSource) {
-            this.testDiscoveryCancellationTokenSource.dispose();
-        }
-        this.testDiscoveryCancellationTokenSource = null;
-        if (disposeTestRunnerToken) {
+    private disposeCancellationToken(tokenType: CancellationTokenType) {
+        if (tokenType === CancellationTokenType.testDicovery) {
+            if (this.testDiscoveryCancellationTokenSource) {
+                this.testDiscoveryCancellationTokenSource.dispose();
+            }
+            this.testDiscoveryCancellationTokenSource = null;
+        } else {
             if (this.testDiscoveryCancellationTokenSource) {
                 this.testDiscoveryCancellationTokenSource.dispose();
             }
@@ -82,7 +89,7 @@ export abstract class BaseTestManager {
         if (isUserInitiated) {
             this.stop();
         }
-        this.createCancellationTokens();
+        this.createCancellationToken(CancellationTokenType.testDicovery);
         return this.discoverTestsPromise = this.discoverTestsImpl(ignoreCache)
             .then(tests => {
                 this.tests = tests;
@@ -105,7 +112,7 @@ export abstract class BaseTestManager {
                     displayTestErrorMessage('There were some errors in disovering unit tests');
                 }
                 storeDiscoveredTests(tests);
-                this.disposeCancellationTokens();
+                this.disposeCancellationToken(CancellationTokenType.testDicovery);
 
                 return tests;
             }).catch(reason => {
@@ -125,7 +132,7 @@ export abstract class BaseTestManager {
                     this.outputChannel.appendLine('' + reason);
                 }
                 storeDiscoveredTests(null);
-                this.disposeCancellationTokens();
+                this.disposeCancellationToken(CancellationTokenType.testDicovery);
                 return Promise.reject(reason);
             });
     }
@@ -165,14 +172,14 @@ export abstract class BaseTestManager {
 
         this._status = TestStatus.Running;
         this.stop();
-        this.createCancellationTokens(true);
+        this.createCancellationToken(CancellationTokenType.testDicovery);
         // If running failed tests, then don't clear the previously build UnitTests
         // If we do so, then we end up re-discovering the unit tests and clearing previously cached list of failed tests
         // Similarly, if running a specific test or test file, don't clear the cache (possible tests have some state information retained)
         const clearDiscoveredTestCache = runFailedTests || moreInfo.Run_Specific_File || moreInfo.Run_Specific_Class || moreInfo.Run_Specific_Function ? false : true;
         return this.discoverTests(clearDiscoveredTestCache, true, true)
             .catch(reason => {
-                if (this.testRunnerCancellationToken && this.testRunnerCancellationToken.isCancellationRequested) {
+                if (this.testDiscoveryCancellationToken && this.testDiscoveryCancellationToken.isCancellationRequested) {
                     return Promise.reject<Tests>(reason);
                 }
                 displayTestErrorMessage('Errors in discovering tests, continuing with tests');
@@ -182,10 +189,11 @@ export abstract class BaseTestManager {
                 };
             })
             .then(tests => {
+                this.createCancellationToken(CancellationTokenType.testRunner);
                 return this.runTestImpl(tests, testsToRun, runFailedTests, debug);
             }).then(() => {
                 this._status = TestStatus.Idle;
-                this.disposeCancellationTokens(true);
+                this.disposeCancellationToken(CancellationTokenType.testRunner);
                 return this.tests;
             }).catch(reason => {
                 if (this.testRunnerCancellationToken && this.testRunnerCancellationToken.isCancellationRequested) {
@@ -195,7 +203,7 @@ export abstract class BaseTestManager {
                 else {
                     this._status = TestStatus.Error;
                 }
-                this.disposeCancellationTokens(true);
+                this.disposeCancellationToken(CancellationTokenType.testRunner);
                 return Promise.reject<Tests>(reason);
             });
     }

--- a/src/client/unittests/common/baseTestManager.ts
+++ b/src/client/unittests/common/baseTestManager.ts
@@ -8,11 +8,17 @@ import { isNotInstalledError } from '../../common/helpers';
 export abstract class BaseTestManager {
     private tests: Tests;
     private _status: TestStatus = TestStatus.Unknown;
-    private cancellationTokenSource: vscode.CancellationTokenSource;
+    private testDiscoveryCancellationTokenSource: vscode.CancellationTokenSource;
+    private testRunnerCancellationTokenSource: vscode.CancellationTokenSource;
     private installer: Installer;
-    protected get cancellationToken(): vscode.CancellationToken {
-        if (this.cancellationTokenSource) {
-            return this.cancellationTokenSource.token;
+    protected get testDiscoveryCancellationToken(): vscode.CancellationToken {
+        if (this.testDiscoveryCancellationTokenSource) {
+            return this.testDiscoveryCancellationTokenSource.token;
+        }
+    }
+    protected get testRunnerCancellationToken(): vscode.CancellationToken {
+        if (this.testRunnerCancellationTokenSource) {
+            return this.testRunnerCancellationTokenSource.token;
         }
     }
     public dispose() {
@@ -21,8 +27,11 @@ export abstract class BaseTestManager {
         return this._status;
     }
     public stop() {
-        if (this.cancellationTokenSource) {
-            this.cancellationTokenSource.cancel();
+        if (this.testDiscoveryCancellationTokenSource) {
+            this.testDiscoveryCancellationTokenSource.cancel();
+        }
+        if (this.testRunnerCancellationTokenSource) {
+            this.testRunnerCancellationTokenSource.cancel();
         }
     }
     constructor(private testProvider: string, private product: Product, protected rootDirectory: string, protected outputChannel: vscode.OutputChannel) {
@@ -40,18 +49,27 @@ export abstract class BaseTestManager {
 
         resetTestResults(this.tests);
     }
-    private createCancellationToken() {
-        this.disposeCancellationToken();
-        this.cancellationTokenSource = new vscode.CancellationTokenSource();
-    }
-    private disposeCancellationToken() {
-        if (this.cancellationTokenSource) {
-            this.cancellationTokenSource.dispose();
+    private createCancellationTokens(createTestRunnerToken: boolean = false) {
+        this.disposeCancellationTokens(createTestRunnerToken);
+        this.testDiscoveryCancellationTokenSource = new vscode.CancellationTokenSource();
+        if (createTestRunnerToken) {
+            this.testRunnerCancellationTokenSource = new vscode.CancellationTokenSource();
         }
-        this.cancellationTokenSource = null;
+    }
+    private disposeCancellationTokens(disposeTestRunnerToken: boolean = false) {
+        if (this.testDiscoveryCancellationTokenSource) {
+            this.testDiscoveryCancellationTokenSource.dispose();
+        }
+        this.testDiscoveryCancellationTokenSource = null;
+        if (disposeTestRunnerToken) {
+            if (this.testDiscoveryCancellationTokenSource) {
+                this.testDiscoveryCancellationTokenSource.dispose();
+            }
+            this.testDiscoveryCancellationTokenSource = null;
+        }
     }
     private discoverTestsPromise: Promise<Tests>;
-    discoverTests(ignoreCache: boolean = false, quietMode: boolean = false): Promise<Tests> {
+    discoverTests(ignoreCache: boolean = false, quietMode: boolean = false, isUserInitiated: boolean = true): Promise<Tests> {
         if (this.discoverTestsPromise) {
             return this.discoverTestsPromise;
         }
@@ -61,8 +79,10 @@ export abstract class BaseTestManager {
             return Promise.resolve(this.tests);
         }
         this._status = TestStatus.Discovering;
-
-        this.createCancellationToken();
+        if (isUserInitiated) {
+            this.stop();
+        }
+        this.createCancellationTokens();
         return this.discoverTestsPromise = this.discoverTestsImpl(ignoreCache)
             .then(tests => {
                 this.tests = tests;
@@ -85,7 +105,7 @@ export abstract class BaseTestManager {
                     displayTestErrorMessage('There were some errors in disovering unit tests');
                 }
                 storeDiscoveredTests(tests);
-                this.disposeCancellationToken();
+                this.disposeCancellationTokens();
 
                 return tests;
             }).catch(reason => {
@@ -95,7 +115,7 @@ export abstract class BaseTestManager {
 
                 this.tests = null;
                 this.discoverTestsPromise = null;
-                if (this.cancellationToken && this.cancellationToken.isCancellationRequested) {
+                if (this.testDiscoveryCancellationToken && this.testDiscoveryCancellationToken.isCancellationRequested) {
                     reason = CANCELLATION_REASON;
                     this._status = TestStatus.Idle;
                 }
@@ -105,7 +125,7 @@ export abstract class BaseTestManager {
                     this.outputChannel.appendLine('' + reason);
                 }
                 storeDiscoveredTests(null);
-                this.disposeCancellationToken();
+                this.disposeCancellationTokens();
                 return Promise.reject(reason);
             });
     }
@@ -144,14 +164,15 @@ export abstract class BaseTestManager {
         }
 
         this._status = TestStatus.Running;
-        this.createCancellationToken();
+        this.stop();
+        this.createCancellationTokens(true);
         // If running failed tests, then don't clear the previously build UnitTests
         // If we do so, then we end up re-discovering the unit tests and clearing previously cached list of failed tests
         // Similarly, if running a specific test or test file, don't clear the cache (possible tests have some state information retained)
         const clearDiscoveredTestCache = runFailedTests || moreInfo.Run_Specific_File || moreInfo.Run_Specific_Class || moreInfo.Run_Specific_Function ? false : true;
-        return this.discoverTests(clearDiscoveredTestCache, true)
+        return this.discoverTests(clearDiscoveredTestCache, true, true)
             .catch(reason => {
-                if (this.cancellationToken && this.cancellationToken.isCancellationRequested) {
+                if (this.testRunnerCancellationToken && this.testRunnerCancellationToken.isCancellationRequested) {
                     return Promise.reject<Tests>(reason);
                 }
                 displayTestErrorMessage('Errors in discovering tests, continuing with tests');
@@ -164,17 +185,17 @@ export abstract class BaseTestManager {
                 return this.runTestImpl(tests, testsToRun, runFailedTests, debug);
             }).then(() => {
                 this._status = TestStatus.Idle;
-                this.disposeCancellationToken();
+                this.disposeCancellationTokens(true);
                 return this.tests;
             }).catch(reason => {
-                if (this.cancellationToken && this.cancellationToken.isCancellationRequested) {
+                if (this.testRunnerCancellationToken && this.testRunnerCancellationToken.isCancellationRequested) {
                     reason = CANCELLATION_REASON;
                     this._status = TestStatus.Idle;
                 }
                 else {
                     this._status = TestStatus.Error;
                 }
-                this.disposeCancellationToken();
+                this.disposeCancellationTokens(true);
                 return Promise.reject<Tests>(reason);
             });
     }

--- a/src/client/unittests/common/baseTestManager.ts
+++ b/src/client/unittests/common/baseTestManager.ts
@@ -69,10 +69,10 @@ export abstract class BaseTestManager {
             }
             this.testDiscoveryCancellationTokenSource = null;
         } else {
-            if (this.testDiscoveryCancellationTokenSource) {
-                this.testDiscoveryCancellationTokenSource.dispose();
+            if (this.testRunnerCancellationTokenSource) {
+                this.testRunnerCancellationTokenSource.dispose();
             }
-            this.testDiscoveryCancellationTokenSource = null;
+            this.testRunnerCancellationTokenSource = null;
         }
     }
     private discoverTestsPromise: Promise<Tests>;

--- a/src/client/unittests/main.ts
+++ b/src/client/unittests/main.ts
@@ -40,7 +40,7 @@ export function activate(context: vscode.ExtensionContext, outputChannel: vscode
     if (settings.unitTest.nosetestsEnabled || settings.unitTest.pyTestEnabled || settings.unitTest.unittestEnabled) {
         // Ignore the exceptions returned
         // This function is invoked via a command which will be invoked else where in the extension
-        discoverTests(true).catch(() => {
+        discoverTests(true, false).catch(() => {
             // Ignore the errors
         });
     }
@@ -61,7 +61,7 @@ async function onDocumentSaved(doc: vscode.TextDocument): Promise<void> {
         return;
     }
 
-    let tests = await testManager.discoverTests(false, true);
+    let tests = await testManager.discoverTests(false, true, false);
     if (!tests || !Array.isArray(tests.testFiles) || tests.testFiles.length === 0) {
         return;
     }
@@ -72,7 +72,7 @@ async function onDocumentSaved(doc: vscode.TextDocument): Promise<void> {
     if (timeoutId) {
         clearTimeout(timeoutId);
     }
-    timeoutId = setTimeout(() => { discoverTests(true); }, 1000);
+    timeoutId = setTimeout(() => { discoverTests(true, false); }, 1000);
 }
 
 function dispose() {
@@ -91,7 +91,7 @@ function registerCommands(): vscode.Disposable[] {
     disposables.push(vscode.commands.registerCommand(constants.Commands.Tests_Discover, () => {
         // Ignore the exceptions returned
         // This command will be invoked else where in the extension
-        discoverTests(true).catch(() => { return null; });
+        discoverTests(true, true).catch(() => { return null; });
     }));
     disposables.push(vscode.commands.registerCommand(constants.Commands.Tests_Run_Failed, () => runTestsImpl(true)));
     disposables.push(vscode.commands.registerCommand(constants.Commands.Tests_Run, (testId) => runTestsImpl(testId)));
@@ -134,7 +134,7 @@ function selectAndRunTestMethod(debug?: boolean) {
     if (!testManager) {
         return displayTestFrameworkError(outChannel);
     }
-    testManager.discoverTests(true, true).then(() => {
+    testManager.discoverTests(true, true, true).then(() => {
         const tests = getDiscoveredTests();
         testDisplay = testDisplay ? testDisplay : new TestDisplay();
         testDisplay.selectTestFunction(getTestWorkingDirectory(), tests).then(testFn => {
@@ -147,7 +147,7 @@ function selectAndRunTestFile() {
     if (!testManager) {
         return displayTestFrameworkError(outChannel);
     }
-    testManager.discoverTests(true, true).then(() => {
+    testManager.discoverTests(true, true, true).then(() => {
         const tests = getDiscoveredTests();
         testDisplay = testDisplay ? testDisplay : new TestDisplay();
         testDisplay.selectTestFile(getTestWorkingDirectory(), tests).then(testFile => {
@@ -164,7 +164,7 @@ function runCurrentTestFile() {
     if (!testManager) {
         return displayTestFrameworkError(outChannel);
     }
-    testManager.discoverTests(true, true).then(() => {
+    testManager.discoverTests(true, true, true).then(() => {
         const tests = getDiscoveredTests();
         const testFiles = tests.testFiles.filter(testFile => {
             return testFile.fullPath === currentFilePath;
@@ -225,7 +225,7 @@ function onConfigChanged() {
 
     // No need to display errors
     if (settings.unitTest.nosetestsEnabled || settings.unitTest.pyTestEnabled || settings.unitTest.unittestEnabled) {
-        discoverTests(true);
+        discoverTests(true, false);
     }
 }
 function getTestRunner() {
@@ -248,7 +248,7 @@ function stopTests() {
         testManager.stop();
     }
 }
-function discoverTests(ignoreCache?: boolean) {
+function discoverTests(ignoreCache?: boolean, isUserInitiated?: boolean) {
     let testManager = getTestRunner();
     if (!testManager) {
         displayTestFrameworkError(outChannel);
@@ -257,7 +257,7 @@ function discoverTests(ignoreCache?: boolean) {
 
     if (testManager && (testManager.status !== TestStatus.Discovering && testManager.status !== TestStatus.Running)) {
         testResultDisplay = testResultDisplay ? testResultDisplay : new TestResultDisplay(outChannel, onDidChange);
-        return testResultDisplay.DisplayDiscoverStatus(testManager.discoverTests(ignoreCache));
+        return testResultDisplay.DisplayDiscoverStatus(testManager.discoverTests(ignoreCache, false, isUserInitiated));
     }
     else {
         return Promise.resolve(null);

--- a/src/client/unittests/nosetest/main.ts
+++ b/src/client/unittests/nosetest/main.ts
@@ -16,7 +16,7 @@ export class TestManager extends BaseTestManager {
     }
     discoverTestsImpl(ignoreCache: boolean): Promise<Tests> {
         let args = settings.unitTest.nosetestArgs.slice(0);
-        return discoverTests(this.rootDirectory, args, this.cancellationToken, ignoreCache, this.outputChannel);
+        return discoverTests(this.rootDirectory, args, this.testDiscoveryCancellationToken, ignoreCache, this.outputChannel);
     }
     runTestImpl(tests: Tests, testsToRun?: TestsToRun, runFailedTests?: boolean, debug?: boolean): Promise<any> {
         let args = settings.unitTest.nosetestArgs.slice(0);
@@ -26,6 +26,6 @@ export class TestManager extends BaseTestManager {
         if (!runFailedTests && args.indexOf('--with-id') === -1) {
             args.push('--with-id');
         }
-        return runTest(this.rootDirectory, tests, args, testsToRun, this.cancellationToken, this.outputChannel, debug);
+        return runTest(this.rootDirectory, tests, args, testsToRun, this.testRunnerCancellationToken, this.outputChannel, debug);
     }
 }

--- a/src/client/unittests/pytest/main.ts
+++ b/src/client/unittests/pytest/main.ts
@@ -14,13 +14,13 @@ export class TestManager extends BaseTestManager {
     }
     discoverTestsImpl(ignoreCache: boolean): Promise<Tests> {
         let args = settings.unitTest.pyTestArgs.slice(0);
-        return discoverTests(this.rootDirectory, args, this.cancellationToken, ignoreCache, this.outputChannel);
+        return discoverTests(this.rootDirectory, args, this.testDiscoveryCancellationToken, ignoreCache, this.outputChannel);
     }
     runTestImpl(tests: Tests, testsToRun?: TestsToRun, runFailedTests?: boolean, debug?: boolean): Promise<any> {
         let args = settings.unitTest.pyTestArgs.slice(0);
         if (runFailedTests === true && args.indexOf('--lf') === -1 && args.indexOf('--last-failed') === -1) {
             args.push('--last-failed');
         }
-        return runTest(this.rootDirectory, tests, args, testsToRun, this.cancellationToken, this.outputChannel, debug);
+        return runTest(this.rootDirectory, tests, args, testsToRun, this.testRunnerCancellationToken, this.outputChannel, debug);
     }
 }

--- a/src/client/unittests/unittest/main.ts
+++ b/src/client/unittests/unittest/main.ts
@@ -16,7 +16,7 @@ export class TestManager extends BaseTestManager {
     }
     discoverTestsImpl(ignoreCache: boolean): Promise<Tests> {
         let args = settings.unitTest.unittestArgs.slice(0);
-        return discoverTests(this.rootDirectory, args, this.cancellationToken, ignoreCache, this.outputChannel);
+        return discoverTests(this.rootDirectory, args, this.testDiscoveryCancellationToken, ignoreCache, this.outputChannel);
     }
     runTestImpl(tests: Tests, testsToRun?: TestsToRun, runFailedTests?: boolean, debug?: boolean): Promise<any> {
         let args = settings.unitTest.unittestArgs.slice(0);
@@ -26,6 +26,6 @@ export class TestManager extends BaseTestManager {
                 return fn.testFunction.status === TestStatus.Error || fn.testFunction.status === TestStatus.Fail;
             }).map(fn => fn.testFunction);
         }
-        return runTest(this, this.rootDirectory, tests, args, testsToRun, this.cancellationToken, this.outputChannel, debug);
+        return runTest(this, this.rootDirectory, tests, args, testsToRun, this.testRunnerCancellationToken, this.outputChannel, debug);
     }
 }

